target_chains/sui/contracts/sources/merkle_tree.move

```diff
@@ -0,0 +1,276 @@
+// Implementation of a Merkle tree in Move. Supports constructing a new tree
+// with a given depth, as well as proving that a leaf node belongs to the tree.
+module pyth::merkle {
+    use std::vector::{Self};
+    use sui::hash::{keccak256};
+    use wormhole::bytes::{Self};
+    use wormhole::bytes20::{Self, Bytes20, data};
+    use wormhole::cursor::{Self, Cursor};
+    use pyth::deserialize::{Self};
+
+    const MERKLE_LEAF_PREFIX: u8 = 0;
+    const MERKLE_NODE_PREFIX: u8 = 1;
+    const MERKLE_EMPTY_LEAF_PREFIX: u8 = 2;
+
+    const E_DEPTH_NOT_LARGE_ENOUGH_FOR_MESSAGES: u64 = 1212121;
+
+    // take keccak256 of input data, then return 20 rightmost bytes of result
+    fun hash(bytes: &vector<u8>): Bytes20 {
+        let hashed_bytes = keccak256(bytes); // 32 byte hash
+        let cursor = cursor::new(hashed_bytes);
+        let _ = bytes::take_bytes(&mut cursor, 12);
+        let bytes20 = bytes::take_bytes(&mut cursor, 20);
+        cursor::destroy_empty(cursor);
+        bytes20::from_bytes(bytes20)
+    }
+
+    fun emptyLeafHash(): Bytes20 {
+        let v = vector::empty();
+        bytes::push_u8(&mut v, MERKLE_EMPTY_LEAF_PREFIX);
+        hash(&v)
+    }
+
+    fun leafHash(data: &vector<u8>): Bytes20 {
+        bytes::push_u8(&mut *data, MERKLE_LEAF_PREFIX);
```

Comment:
shouldn't we push_front here? :? it seems push_u8 pushes to the back

Reply:
it seems dangerous for this function to mutate data, as if someone wants to use it later, then they may be surprised.

Reply:
data is passed in as a read-only reference, so the original copy is not mutated

Reply:
you guys are right about how push_u8 pushes back

Reply:
I believe it would have to be `data: &mut vector<u8>` for there to be a danger of mutating the original data, otherwise it makes a copy inside this function